atlas/responder/pkg/svc/zserver.go

```diff
@@ -0,0 +1,341 @@
+package svc
+
+import (
+	models "atlas/responder/internal"
+	"atlas/responder/pkg/pb"
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	dapr "github.com/dapr/dapr/pkg/proto/runtime/v1"
+	"github.com/dapr/go-sdk/service/common"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// version is the current version of the service
+	version            = "0.0.1"
+	invalidServiceName = "invalid service name"
+	emptyRequest       = "empty Request"
+	success            = "success"
+	errorMsg           = "error"
+	errorMissingResp   = "missing repository response error"
+	hiddenUptimeMsg    = "uptime is hidden, mode = false"
+	pubTopic           = "neworder"
+	subTopic           = "neworder2"
+	route              = "/orders2"
+	portal             = "portal"
+	responder          = "responder"
+	storage            = "storage"
+)
+
+type Responder struct {
+	pb.ResponderServer
+	responses *sync.Map
+	logger    *logrus.Logger
+	responder models.Service
+	client    dapr.DaprClient
+	Sub       *common.Subscription
+}
+
+func getResponder() models.Service {
+	return models.Service{
+		ServiceName:          "responder",
+		ServiceDesc:          "responder service desc",
+		ServiceUptime:        time.Now(),
+		ServiceCountRequests: 0,
+	}
+}
+
+func NewResponder(logger *logrus.Logger, s *sync.Map) (*Responder, error) {
+	sub := &common.Subscription{
+		PubsubName: "messages",
+		Topic:      subTopic,
+		Route:      route,
+	}
+
+	conn, err := grpc.Dial(fmt.Sprintf("0.0.0.0:%s", os.Getenv("DAPR_GRPC_PORT")), grpc.WithInsecure())
+	if err != nil {
+		return nil, fmt.Errorf("failed to open atlas pubsub connection: %v", err)
+	}
+
+	return &Responder{
+		logger:    logger,
+		responses: s,
+		responder: getResponder(),
+		client:    dapr.NewDaprClient(conn),
+		Sub:       sub,
+	}, nil
+}
+
+func (a *Responder) GetVersion(ctx context.Context, in *empty.Empty) (*pb.VersionResponse, error) {
+	return &pb.VersionResponse{Version: version}, nil
+}
+
+func (a *Responder) GetInfo(ctx context.Context, in *pb.GetInfoRequest) (*pb.GetInfoResponse, error) {
+	if in == nil || in.Service == "" {
+		return nil, errors.New(emptyRequest)
+	}
+
+	if in.Service == responder {
+		a.responder.ServiceCountRequests++
+		return &pb.GetInfoResponse{Value: a.responder.ServiceDesc}, nil
+	}
+
+	if in.Service == storage {
+		id := getId(time.Now().String())
+		if err := a.PublishMsg(ctx, id, "getInfo", nil); err != nil {
+			return nil, err
+		}
+		time.Sleep(time.Millisecond * 15)
```

Comment:
Here and below: why do we need to sleep for 15ms? Dapr works asynchronously, and you have absolutely no warranties, when the remote side will respond to your request. That means, you may get an answer in 1ms, or in 10 hours, or even never. This limitation should be solved in other way, without adding the artificial delays.

Reply:
Done, i use sync.Map with channel instead of time.Sleep